### PR TITLE
fix(jsonpath): Add JsonPath grammar for the child identifier in brackets

### LINKED ISSUE #3511: FT.SEARCH - search on TAG not working
**Describe the bug**
When performing a FT.SEARCH query on a TAG field on a JSON document, no results are returned. However, the same functionality works both on a HASH document in dragonfly and a JSON document using redisearch

**To Reproduce**
1. Create the index:
```
FT.CREATE "json_city" ON JSON PREFIX 1 json_city: SCHEMA "$[\"name\"]" AS name TAG SEPARATOR | "$[\"population\"]" AS population NUMERIC SORTABLE "$[\"continent\"]" AS continent TAG SEPARATOR |
```
2. Create the records:
```
JSON.SET "json_city:01J57VZMXMTXNA912J4BV2NFV8" $ "{\"name\":\"London\",\"population\":8.8,\"continent\":\"Europe\"}"
JSON.SET "json_city:01J57VZMXZEF6WFQ5CS29JHP4F" $ "{\"name\":\"Athens\",\"population\":3.1,\"continent\":\"Europe\"}"
JSON.SET "json_city:01J57VZMXZXMS9EZ1QFH0YFS24" $ "{\"name\":\"Tel-Aviv\",\"population\":1.3,\"continent\":\"Asia\"}"
JSON.SET "json_city:01J57VZMY0Z1VY8WH7X784CFA8" $ "{\"name\":\"Hyderabad\",\"population\":9.8,\"continent\":\"Asia\"}"
```
3. Query all records to ensure index is setup correctly: 
```
FT.SEARCH json_city *
```
4. Search for cities that have `Europe` as the `TAG` attribute `continent`.
```
FT.SEARCH json_city "(@continent:{Europe})"
```
5. Observe no results found
```
127.0.0.1:6379> FT.SEARCH json_city "(@continent:{Europe})"
1) (integer) 0
```

**Expected behavior**
Expected behavior would be that cities with the `continent` field matching `Europe` would be found, like they are when using hashes or redisearch. This behavior is outlined in the [dragonfly FT.SEARCH docs](https://www.dragonflydb.io/docs/command-reference/search/ft.search).

**Screenshots**
dragonfly (not working as expected):
image
[dragonfly docs](https://www.dragonflydb.io/docs/command-reference/search/ft.search):
Screenshot 2024-08-14 at 2 56 23 AM
dragonfly (working example using hashes) - taken from [dragonfly search announcement example](https://www.dragonflydb.io/blog/announcing-dragonfly-search):
image
redis-stack (uses exact same code as reproduction steps above and demonstrates expected behavior)
Screenshot 2024-08-14 at 3 05 15 AM
Screenshot 2024-08-14 at 3 05 25 AM
Screenshot 2024-08-14 at 3 05 31 AM
Screenshot 2024-08-14 at 3 05 37 AM

**Environment (please complete the following information):**
 - OS: Mac OS 14.6.1 (23G93)
 - Kernel: Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:30 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6000 arm64
 - Containerized?: Docker/Docker compose
 - Dragonfly Version: df-v1.21.1

**Reproducible Code Snippet**
```
FT.CREATE "json_city" ON JSON PREFIX 1 json_city: SCHEMA "$[\"name\"]" AS name TAG SEPARATOR | "$[\"population\"]" AS population NUMERIC SORTABLE "$[\"continent\"]" AS continent TAG SEPARATOR |

JSON.SET "json_city:01J57VZMXMTXNA912J4BV2NFV8" $ "{\"name\":\"London\",\"population\":8.8,\"continent\":\"Europe\"}"
JSON.SET "json_city:01J57VZMXZEF6WFQ5CS29JHP4F" $ "{\"name\":\"Athens\",\"population\":3.1,\"continent\":\"Europe\"}"
JSON.SET "json_city:01J57VZMXZXMS9EZ1QFH0YFS24" $ "{\"name\":\"Tel-Aviv\",\"population\":1.3,\"continent\":\"Asia\"}"
JSON.SET "json_city:01J57VZMY0Z1VY8WH7X784CFA8" $ "{\"name\":\"Hyderabad\",\"population\":9.8,\"continent\":\"Asia\"}"

FT.SEARCH json_city *

FT.SEARCH json_city "(@continent:{Europe})"
```

**Additional context**
Code for working hash example:
```
FT.CREATE cities PREFIX 1 city: SCHEMA name TEXT population NUMERIC SORTABLE continent TAG
HSET city:1 name London population 8.8 continent Europe
HSET city:2 name Athens population 3.1 continent Europe
HSET city:3 name Tel-Aviv population 1.3 continent Asia
HSET city:4 name Hyderabad population 9.8 continent Asia
FT.INFO cities
FT.SEARCH cities '@continent:{Europe}' SORTBY population DESC LIMIT 0 1 RETURN 2 name population
FT.SEARCH cities '@population:[0 5] @continent:{Asia}' RETURN 1 name
```

I've also tried escaping the string in many different ways, as some other issues mention incorrect query formatting. as dragonfly is intended to be a drop in redis replacement i would hope that i wouldn't need to format the query in a different way to get it to work.
```
FT.SEARCH json_city "(@continent:{Europe})"
FT.SEARCH json_city (@continent:{Europe})
FT.SEARCH json_city "@continent:\{Europe\}"
FT.SEARCH json_city "@continent:\{'Europe'\}"
```

My ultimate goal is to get this type of query working with [redis-om-node](https://github.com/redis/redis-om-node)

Thanks for your help!

### PR DESCRIPTION
**Pull request by BagritsevichStepan**
fixes #3511

### CODE REVIEW

File: src/core/json/jsonpath_grammar.y [hunk 1]
```diff
@@ -76,21 +80,28 @@ opt_relative_location:
...
-
+        | bracket_expr
 
 identifier: UNQ_STR
-         // | single_quoted_string | double_quoted_string
 
-bracket_index: WILDCARD { $$ = PathSegment{SegmentType::INDEX, IndexExpr::All()}; }
+bracket_expr: LBRACKET bracket_key RBRACKET { driver->AddSegment($2); } opt_relative_location
+
+bracket_key: single_quoted_string { $$ = PathSegment(SegmentType::IDENTIFIER, $1); }
```
Comment (BagritsevichStepan) [inline src/core/json/jsonpath_grammar.y:93]: I'm not sure that `bracket_key` is the best name for this nonterminal. Open for suggestions

**Comment by dranikpg** [comment]
Might fix #3145 as well 👍🏻 

Can you please explain what the cause was? I.e. was it that jsonv1 correctly processes quotes internally?

File: src/core/json/jsonpath_grammar.y (hunk 1 above)
Comment (dranikpg) [reply src/core/json/jsonpath_grammar.y:93]: Bracket index was an ok name, both strings and integers can be "index"... Suggestions would be bracket_inner, bracket_core 🤷🏻‍♂️

File: src/server/json_family_test.cc [hunk 2]
```diff
@@ -223,6 +223,52 @@ TEST_F(JsonFamilyTest, SetGetFromPhonebook) {
       R"([st{stt"number":s"212 555-1234",stt"type":s"home"st},st{stt"number":s"646 555-4567",stt"type":s"office"st}s])");
 }
 
+TEST_F(JsonFamilyTest, GetBrackets) {
```
Comment (dranikpg) [inline src/server/json_family_test.cc:226]: Good that you added a lot of tests for basic behaviour 👍🏻 

I'd maybe add an intermediate one, for example a case where brackets are chained, i.e. ["a"]['b']["c"] or ["a"].b['c'], etc (if it's even allowed to mix it [] and . 🙂)

**Review by dranikpg — APPROVED** [review]

**Review by dranikpg — APPROVED** [review]
👍🏻👍🏻👍🏻

**Review by romange — APPROVED** [review]
Good work @BagritsevichStepan !

**Pull request merged**